### PR TITLE
Feature/HDF5-Output-SoA

### PR DESCRIPTION
**Pull request by jpmeng**
This pull request is for two capabilities:

1.  Writing out data as slab or plane in a lower precision using HDF5 (e.g.,  running in double precision but writing out in single or half precision 

2.  In the SoA mode, setting the ops_dat memory from a buffer in AoS mode and writing out ops_dat in AoS layout.

### CODE REVIEW

File: ops/c/include/ops_hdf5_common.h [hunk 1]
```diff
@@ -38,7 +38,7 @@
 
 #ifndef __OPS_HDF5_COMMON_H
 #define __OPS_HDF5_COMMON_H
-#include "hdf5.h"
+#include  "hdf5.h"
```
Comment (reguly) [inline ops/c/include/ops_hdf5_common.h:41]: Please remove extra whitespace

File: ops/c/include/ops_util.h [hunk 2]
```diff
@@ -110,6 +109,7 @@ inline int off2D(int dim, int *start, int *end, int *size, int *stride) {
   return off;
 }
 
+
```
Comment (reguly) [inline ops/c/include/ops_util.h:112]: Remove extra line

File: ops/c/src/core/ops_dummy_singlenode.cpp [hunk 3]
```diff
@@ -626,88 +551,97 @@ void ops_dat_fetch_data_host(ops_dat dat, int part, char *data) {
...
-  ops_dat_get_extents(dat, part, ldisp, lsize);
-  for (int d = dat->block->dims; d < OPS_MAX_DIM; d++) {
-    lsize[d] = 1;
-    ldisp[d] = 0;
+
+  int *range{new int(2 * dat->block->dims)};
+  for (int d = 0; d < dat->block->dims; d++) {
+    range[2 * d] = dat->d_m[d];
+    range[2 * d + 1] = dat->size[d] - dat->d_p[d];
+    // ops_printf("range[%d]=%d range[%d]=%d size=%d d_p=%d d_m=%d\n", 2 * d,
```
Comment (reguly) [inline ops/c/src/core/ops_dummy_singlenode.cpp:612]: Please remove commented out code

File: ops/c/src/core/ops_dummy_singlenode.cpp [hunk 4]
```diff
@@ -626,88 +551,97 @@ void ops_dat_fetch_data_host(ops_dat dat, int part, char *data) {
...
-  int ldisp[OPS_MAX_DIM] = {0};
-  ops_dat_get_extents(dat, part, ldisp, lsize);
-  for (int d = dat->block->dims; d < OPS_MAX_DIM; d++) {
-    lsize[d] = 1;
-    ldisp[d] = 0;
+
+  int *range{new int(2 * dat->block->dims)};
+  for (int d = 0; d < dat->block->dims; d++) {
+    range[2 * d] = dat->d_m[d];
+    range[2 * d + 1] = dat->size[d] - dat->d_p[d];
```
Comment (reguly) [inline ops/c/src/core/ops_dummy_singlenode.cpp:611]: I am not sure this range calculation is right - dat->size[d] is compute as user_given_size[d]-d_m[d]+d_p[d] what is the intent?

File: ops/c/src/externlib/ops_hdf5_common.cpp [hunk 5]
```diff
@@ -82,4 +82,165 @@ void split_h5_name(const char *data_name,
...
+  if (H5Lexists(parent_group, data_name, H5P_DEFAULT) == 0) {
+    hid_t data_plist_id{H5Pcreate(H5P_DATASET_CREATE)};
+    file_space = H5Screate_simple(data_dims, global_data_size, NULL);
+
+    hid_t type = h5_type(data_type);
+    if (type == H5T_NATIVE_FLOAT || type == H5T_NATIVE_DOUBLE) {
+      hid_t real_type;
+      bool float16_created{false};
+      if (type == H5T_NATIVE_DOUBLE) {
+        if (real_precision == 16) {
```
Comment (reguly) [inline ops/c/src/externlib/ops_hdf5_common.cpp:139]: what is 4, 8, 16 here? 4 is half precision? that is very misleading, because sizeof(float) = 4

**Review by reguly — CHANGES_REQUESTED** [review]

File: ops/c/src/core/ops_dummy_singlenode.cpp (hunk 3 above)
Comment (jpmeng) [reply ops/c/src/core/ops_dummy_singlenode.cpp:612]: Done

File: ops/c/include/ops_hdf5_common.h (hunk 1 above)
Comment (jpmeng) [reply ops/c/include/ops_hdf5_common.h:41]: Done

File: ops/c/include/ops_util.h (hunk 2 above)
Comment (jpmeng) [reply ops/c/include/ops_util.h:112]: Done

File: ops/c/src/externlib/ops_hdf5_common.cpp (hunk 5 above)
Comment (jpmeng) [reply ops/c/src/externlib/ops_hdf5_common.cpp:139]: They are intended for useful figures of a real number at various precision. A kind of Fortran way adopted (i.e., selected_real_kind) here. Happy to move to other conventions, let's do 2 for half, 4 for float and 8 for double?

Comment (jpmeng) [reply ops/c/src/externlib/ops_hdf5_common.cpp:139]: The HDF5 convention is 16(half) 32(single) 64(double), we can probably use as well

Comment (jpmeng) [reply ops/c/src/externlib/ops_hdf5_common.cpp:139]: We might define some  enumeration for this.

File: ops/c/src/core/ops_dummy_singlenode.cpp (hunk 4 above)
Comment (reguly) [reply ops/c/src/core/ops_dummy_singlenode.cpp:611]: in that case, the second line should be dat->size[d] + dat->d_m[d]; Please make sure this is consistent with the MPI version - but there the intra-block MPI halos are of course not included

File: ops/c/src/externlib/ops_hdf5_common.cpp (hunk 5 above)
Comment (reguly) [reply ops/c/src/externlib/ops_hdf5_common.cpp:139]: either bits - then it's 16,32,64, or bytes - 2,4,8. Yes, and enum would be good for this!

**Comment by reguly** [comment]
Code looks OK. Final request: please create some tests for this app, mainly comparing the output of MPI and non-MPI versions to make sure they match.

**Comment by jpmeng** [comment]
> Code looks OK. Final request: please create some tests for this app, mainly comparing the output of MPI and non-MPI versions to make sure they match.

What did you mean? Testing the output under SoA mode but comparing those of MPI and non-MPI mode?

**Comment by reguly** [comment]
Yes, that’s right.

**Comment by jpmeng** [comment]
Sure. Tests have been added. Also, all tests will run when merging to the develop branch now. We probably need to allow a pull to develop to see if all tests can be past though they seem Ok on my laptop.

**Comment by reguly** [comment]
Should we test the hdf5 slice app too the same way for consistency between MPI and non-MPI backends?

**Comment by jpmeng** [comment]
> Should we test the hdf5 slice app too the same way for consistency between MPI and non-MPI backend
For this particular branch of IO under the SoA mode, I think the multiDim-HDF5 test is pretty enough.  hdf5_slice itself might need some tests for the related capabilties. For this purpuse, however there is another capablity to come soon, i.e., stride wrtiting. Probably better to add the test there.

**Review by reguly — APPROVED** [review]

**Pull request merged**